aspnetcore/fundamentals/static-files/samples/8.x/StaticFileAuth/Program.cs

```diff
@@ -0,0 +1,201 @@
+            });
+            break;
+        default:
+            return Results.Unauthorized();
+    }
+
+    var jwtTokenHandler = new JwtSecurityTokenHandler();
+    
+    return Results.Ok(new {
+        access_token = jwtTokenHandler.WriteToken(jwtTokenHandler.CreateToken(tokenDescriptor))
+    });
+}).AllowAnonymous();
+
+string GetOrCreateFilePath(string nestedFilesDirectory, string fileName, string rootFilesDirectory = "PrivateFiles")
+{
+    var rootFilesDirectoryPath = Path.Combine(app.Environment.ContentRootPath, rootFilesDirectory);
+    
+    if (!Directory.Exists(rootFilesDirectoryPath))
+        Directory.CreateDirectory(rootFilesDirectoryPath);
+    
+    var nestedFilesDirectoryPath = Path.Combine(app.Environment.ContentRootPath, rootFilesDirectory, nestedFilesDirectory);
+    
+    if (!Directory.Exists(nestedFilesDirectoryPath))
+        Directory.CreateDirectory(nestedFilesDirectoryPath);
+
+    return Path.Combine(app.Environment.ContentRootPath, rootFilesDirectory, nestedFilesDirectory, fileName);
+}
+
+async Task SaveFileWithCustomFileName(IFormFile file, string directoryAlias, string fileSaveName)
+{
+    var filePath = GetOrCreateFilePath(directoryAlias, fileSaveName);
+    await using var fileStream = new FileStream(filePath, FileMode.Create);
+    await file.CopyToAsync(fileStream);
+}
+
+app.MapGet("/files/{fileName}", (string fileName, HttpContext context) =>
+    {
+        string filePath;
+        
+        if (context.User.Claims.Any(u => u.Subject.HasClaim("isAdmin", "true")))
+        {
+            filePath = GetOrCreateFilePath("Admin", fileName);
+        }
+        else
+        {
+            var alias = context.User.FindFirstValue("alias");
+
+            if (string.IsNullOrEmpty(alias))
+                return Results.Unauthorized();
+
+            filePath = GetOrCreateFilePath(alias, fileName);
+        }
+        
+        if (File.Exists(filePath))
+            return TypedResults.PhysicalFile(filePath, fileDownloadName: $"{fileName}");
+
+        return TypedResults.NotFound("No file found with the supplied file name");
+    })
+    .WithName("GetFileByName")
+    .RequireAuthorization("PrivateFiles");
+
+app.MapPost("/files", async (IFormFile file, LinkGenerator linker, HttpContext context) =>
+    {
+        if (!Utilities.IsFileValid(file))
+            return Results.ValidationProblem(new Dictionary<string, string[]>
+            {
+                {
+                    "invalid.extensions", new[]
+                    {
+                        $"Only allowed file types are [{string.Join(", ", Utilities.AllowedFileSignatures.Keys)}]"
+                    }
+                }
+            });
+
+        var fileSaveName = Guid.NewGuid().ToString("N") + Path.GetExtension(file.FileName);
+        
+        if (context.User.Claims.Any(u => u.Subject.HasClaim("isAdmin", "true")))
+            await SaveFileWithCustomFileName(file, "Admin", fileSaveName);
```

Comment:
This snippet looks for the `isAdmin` claim for an admin user. If it is not there, it looks for the `alias` claim for regular users. We can't return something from here. On another thought, I'm looking for some advice. Is it a bit complicated for the reader to understand that we are looking for two separate claims in the token?

Reply:
SaveFileWithCustomFileName is getting called twice if the user is an admin. You at least need an else block to avoid the second call.

The claims approach is a bit muddled. How about a simpler model where:
A) You only need to be an authenticated user to download.
B) You need to be an authenticated user with the admin claim to upload.

Then you don't have separate directories, just one folder to work with.